lxd/api_project.go

```diff
@@ -968,8 +968,11 @@ func projectIsEmpty(ctx context.Context, project *cluster.Project, tx *db.Cluste
 		return false, err
 	}
 
-	if len(images) > 0 {
-		return false, nil
+	// Check for images that were not automatically cached.
```

Comment:
Did you consult where `projectIsEmpty` is used (other than when deleting a project)?

As a cursory glance shows me its also used in projectPost just before allowing a project rename?

Does this change potentially allow images to become orphaned if a project is renamed?

Also for project deletion, does this change require any additional changes to ensure that cached images that previously needed to be manually removed before project deletion are now deleted automatically when the project is deleted?

Reply:
Hi @tomponline, thanks for the comment.

> Did you consult where projectIsEmpty is used (other than when deleting a project)?
> As a cursory glance shows me its also used in projectPost just before allowing a project rename?
> Does this change potentially allow images to become orphaned if a project is renamed?

Yes, currently just used when renaming and deleting. Tested renaming while cached images are present, and the images are not orphaned after rename. To avoid potential issues in the future though, I've added a boolean parameter `ignoreCachedImages` to `projectIsEmpty` so that the callee can make the decision.
```
$ lxc project switch foo
$ lxc image list
+-------+--------------+--------+---------------------------------------------+--------------+-----------+-----------+------------------------------+
| ALIAS | FINGERPRINT  | PUBLIC |                 DESCRIPTION                 | ARCHITECTURE |   TYPE    |   SIZE    |         UPLOAD DATE          |
+-------+--------------+--------+---------------------------------------------+--------------+-----------+-----------+------------------------------+
|       | 74957a558028 | no     | ubuntu 24.04 LTS amd64 (release) (20241004) | x86_64       | CONTAINER | 241.51MiB | Oct 22, 2024 at 7:31pm (UTC) |
+-------+--------------+--------+---------------------------------------------+--------------+-----------+-----------+------------------------------+
$ lxc image info 74957a558028
Fingerprint: 74957a5580288913be8a8727d121f16616805e3183629133029ca907f210f541
Size: 241.51MiB
Architecture: x86_64
Type: container
Public: no
Timestamps:
    Created: 2024/10/04 00:00 UTC
    Uploaded: 2024/10/22 19:31 UTC
    Expires: 2029/05/31 00:00 UTC
    Last used: 2024/10/22 19:31 UTC
Properties:
    release: noble
    serial: 20241004
    type: squashfs
    version: 24.04
    architecture: amd64
    description: ubuntu 24.04 LTS amd64 (release) (20241004)
    label: release
    os: ubuntu
Aliases:
Cached: yes
Auto update: enabled
Source:
    Server: https://cloud-images.ubuntu.com/releases
    Protocol: simplestreams
    Alias: 24.04
Profiles:
    - default
$ lxc project switch default
$ lxc project rename foo bar
Project foo renamed to bar
$ lxc project switch bar
$ lxc image ls
+-------+--------------+--------+---------------------------------------------+--------------+-----------+-----------+------------------------------+
| ALIAS | FINGERPRINT  | PUBLIC |                 DESCRIPTION                 | ARCHITECTURE |   TYPE    |   SIZE    |         UPLOAD DATE          |
+-------+--------------+--------+---------------------------------------------+--------------+-----------+-----------+------------------------------+
|       | 74957a558028 | no     | ubuntu 24.04 LTS amd64 (release) (20241004) | x86_64       | CONTAINER | 241.51MiB | Oct 22, 2024 at 7:31pm (UTC) |
+-------+--------------+--------+---------------------------------------------+--------------+-----------+-----------+------------------------------+
```

> Also for project deletion, does this change require any additional changes to ensure that cached images that previously needed to be manually removed before project deletion are now deleted automatically when the project is deleted?

Thanks for catching this. I've updated the PR to include code which deletes the cached images before deleting the project, if they exist.